call.go

```diff
@@ -130,3 +133,170 @@ func WithRoutingKey(rk string) CallOption {
 func WithRoutingDelegate(rd string) CallOption {
 	return CallOption{func(o *OutboundCall) { o.routingDelegate = &rd }}
 }
+
+// InboundCall is an incoming call. It holds information about the inbound
+// call and its response.
+//
+// Encoding authors may use InboundCall to provide information about the
+// incoming request on the Context and receive response headers through
+// WriteResponseHeader.
+type InboundCall struct {
+	resHeaders []keyValuePair
+	req        *transport.Request
+}
+
+type inboundCallKey struct{} // context key for *InboundCall
+
+// NewInboundCall builds a new InboundCall with the given context.
+//
+// A request context is returned and must be used in place of the original.
+func NewInboundCall(ctx context.Context) (context.Context, *InboundCall) {
+	call := &InboundCall{}
+	return context.WithValue(ctx, inboundCallKey{}, call), call
+}
+
+// getInboundCall returns the inbound call on this context or nil.
+func getInboundCall(ctx context.Context) (*InboundCall, bool) {
+	call, ok := ctx.Value(inboundCallKey{}).(*InboundCall)
+	return call, ok
+}
+
+// ReadFromRequest reads information from the given request.
+//
+// This information may be queried on the context using functions like Caller,
+// Service, Procedure, etc.
+func (ic *InboundCall) ReadFromRequest(req *transport.Request) error {
+	// TODO(abg): Maybe we should copy attributes over so that changes to the
+	// Request don't change the output.
+	ic.req = req
+	return nil
+}
+
+// WriteToResponse writes response information from the InboundCall onto the
+// given ResponseWriter.
+//
+// If used, this must be called before writing the response body to the
+// ResponseWriter.
+func (ic *InboundCall) WriteToResponse(resw transport.ResponseWriter) error {
```

Comment:
It's not automatic. The encoding must call this explicitly to write response headers for non-oneway responses.